### PR TITLE
feat(ui): review item component

### LINKED ISSUE #51: [Feat][UI] ReviewItem component
## Mockups / Screenshots

ReviewItem component 

image

https://www.figma.com/file/nSCkMgGjpXTgLFEHykk0yb/Product-Design?type=design&node-id=324%3A20715&mode=dev

LockCta Component

image

https://www.figma.com/file/nSCkMgGjpXTgLFEHykk0yb/Product-Design?type=design&node-id=324%3A23707&mode=dev

## Implementation

ReviewItem component 
- `/common/components/ReviewItem`
- Create `ReviewItem.theme.ts` to manage variants and size. Refer to #19 for guide
- Use the other components that are already available as much as possible

LockCta Component
- `/common/components/LockCta`
- Create `LockCta.theme.ts` to manage variants and size.
- Use the other components that are already available as much as possible

### PR DESCRIPTION
**Pull request by davidlhw**
> ‼️ Note: This PR is branched on top of [feature/tag-component](https://github.com/AfterClass-io/afterclass.io-v2/tree/feature/tag-component) and should be merged only after #63 is merged

closes #51

## changes

- \+ `<ReviewItem />` component & it's subcomponents under `@/common/components/ReviewItem/`
- \+ new button variant `rounded`
- update `<LockCtaOverlay />` component to have a new variant for reviews _(previously was only made for rating section)_
- \+ `<ThumbsUpFilledIcon />` component
- \+ `index.ts` for functions _(every file in the `@/common/functions` dir should only house and export 1 function)_

## implementation

- the "Show more" button doesnt do anything at this point as the modal component for the full review view is not yet ready

## preview

image

image

### CODE REVIEW

File: src/common/components/ReviewItem/ReviewItem.tsx [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+  const HeaderReviewer = () => (
+    <div className={profileContainer()}>
+      <div className={profileIcon()}>
+        <div className="h-4 w-4 rounded-full bg-cyan-800"></div>
+      </div>
+      <div className={profileName()}>{review.username}</div>
+    </div>
+  );
+
+  const ReviewContent = ({ isDetailed }: { isDetailed: boolean }) => {
```
Comment (Aztriltus) [inline src/common/components/ReviewItem/ReviewItem.tsx:68]: I think we can move this as its own component under `ReviewItem/ReviewContent`

Comment (Aztriltus) [reply src/common/components/ReviewItem/ReviewItem.tsx:68]: It is generally not advisable to declare components within a component i.e. `ReviewContent` inside of `ReviewItem` due it re-rendering whenever ReviewItem changes.

Maybe all the components like `HeaderReviewer` can be renamed and moved to a subfolder like `ReviewItem/ReviewerProfile`. Wdyt?

**Review by Aztriltus — COMMENTED** [review]
Some comments!

File: src/common/components/ReviewItem/ReviewItem.tsx [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+  username: string;
+  likeCount: number;
+  labels: Label[];
+  createdAt: number;
+};
+
+export type ReviewItemProps = ReviewItemVariants & {
+  review: Review;
+  isLocked?: boolean;
+  variant?: "home" | "subpage";
```
[the remark this answers is not in the document]
Comment (DavidLHW) [reply src/common/components/ReviewItem/ReviewItem.tsx]: great point! I noticed that the design had 3 different variants:

1. 'home' - where the left profile is the course and the right profile is the user
2. 'professor' - where the left profile is the user and the right profile is the course
3. 'course' - where the left profile is the user and right profile is hidden

I've updated the variants to reflect this difference at [`2eaf9f01`](https://github.com/AfterClass-io/afterclass.io-v2/pull/64/commits/2eaf9f01aaaf5deb5d527482bd985b2579610917#diff-784660bcacaed5171ae3838a6ae0a17385e4a7d322cb6d498541c3d5f6137606R28)

File: src/common/components/ReviewItem/ReviewItem.tsx (hunk 1 above)
Comment (DavidLHW) [reply src/common/components/ReviewItem/ReviewItem.tsx]: updated at 2eaf9f01!

File: src/common/components/ReviewItem/ReviewItem.tsx [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+          <ProfileReviewer name={review.username} />
+        )}
+        <div className={metadataContainer()}>
+          {variant === "home" ? (
+            <ProfileReviewer name={review.username} />
+          ) : (
+            variant === "professor" && (
+              <ProfileSchool courseCode={review.courseCode} />
+            )
+          )}
```
Comment (Aztriltus) [inline src/common/components/ReviewItem/ReviewItem.tsx:60]: I think ternary condition might not be easy to read. How about something like this?

```suggestion
          {variant === "home" && (
            <ProfileReviewer name={review.username} />
          )}
           {variant === "professor" && (
              <ProfileSchool courseCode={review.courseCode} />
            )}
```

File: src/common/components/ReviewItem/ReviewItem.tsx [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            iconRight={<ThumbUpFilledIcon className={likeIcon()} />}
+          >
+            {review.likeCount}
+          </Button>
+          <div className={timedelta()}>
+            {getHumanReadableTimestampDelta(review.createdAt)}
+          </div>
+        </div>
+      </div>
+      {isLocked ? (
```
Comment (Aztriltus) [inline src/common/components/ReviewItem/ReviewItem.tsx:73]: This ternary is fine though!

File: src/common/components/ReviewItem/ReviewItem.tsx [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+          ) : (
+            variant === "professor" && (
+              <ProfileSchool courseCode={review.courseCode} />
+            )
+          )}
+          <Button
+            rounded
+            variant="secondary"
+            iconRight={<ThumbUpFilledIcon className={likeIcon()} />}
+          >
```
Comment (Aztriltus) [inline src/common/components/ReviewItem/ReviewItem.tsx:65]: Maybe we can use a smaller size via the `size` prop

File: src/common/components/Button/Button.theme.ts [hunk 6]
```diff
@@ -127,6 +127,18 @@ export const buttonTheme = tv(
...
+      rounded: {
+        true: [
+          "inline-flex",
+          "min-w-14",
+          "py-[0.125rem]",
+          "px-3",
+          "gap-2",
+          "rounded-[6.1875rem]",
+          "border-solid",
+          "border-border-default",
```
Comment (Aztriltus) [inline src/common/components/Button/Button.theme.ts:139]: Somehow the hover bg doesn't look as rounded as the default state 

https://github.com/AfterClass-io/afterclass.io-v2/assets/6003064/7b861157-4be0-48b9-987a-f1e7ef291891

**Review by Aztriltus — COMMENTED** [review]
Some comments!

File: src/common/components/ReviewItem/ReviewItem.tsx (hunk 3 above)
Comment (davidlhw) [reply src/common/components/ReviewItem/ReviewItem.tsx]: sounds good! accepted suggestion at 5d75c58

File: src/common/components/Button/Button.theme.ts (hunk 6 above)
Comment (davidlhw) [reply src/common/components/Button/Button.theme.ts]: great catch! fixed at f4f6aec

**Review by Aztriltus — APPROVED** [review]
Nice, LGTM. Thanks for the changes!

**Pull request merged**